app.go

```diff
@@ -138,7 +138,13 @@ type App struct {
 
 // New creates and initializes an App. All applications begin with the
 // Lifecycle type available in their dependency injection container.
+//
+// It then executes all functions supplied via the Invoke option. Supplying
+// arguments to these functions requires calling some of the constructors
+// supplied by the Provide option. If any invoked function fails, an error is
+// returned immediately.
 func New(opts ...Option) *App {
+	// TODO(glib): New should probably return (*App, error)
```

Comment:
Think of a panic as a marginally more polite segfault; just because there's nothing that the user can do to fix this app object doesn't mean that we shouldn't shut the process down gracefully. At the very least, any other long-lived goroutines should have the chance to drain, checkpoint their state, and exit. `Run` gets to cheat, but `New`, `Start`, `Done`, and `Stop` should behave themselves and let the service owner decide how to handle errors.

I'm not opposed to returning an error from `New`, but we don't necessarily need to.

We can instead add an `Err` method that returns `app.optionErr` (which we should rename to `app.err`). This is the same approach used to keep [`bufio.Scanner`](https://godoc.org/bufio#Scanner) easy to use without resorting to panics. It should let 

```
fx.New(...some options...).Run()
```
continue to work, since `Start` can continue to short-circuit if `New` failed.

Users who go the fully-manual route have to handle the error somewhere, so they may end up with

```
app := fx.New(...options...)
if err := app.Err(); err != nil {
  log.Fatalf(err)
}
...usual context creation/Start/Done/Stop dance...
```
If they're just going to call `Start` right away, though, they can skip the explicit `Err` check.